<a href="https://colab.research.google.com/github/payasvi/dl/blob/master/prac8/Keras_losses_DL_Prac_8(iii).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Usage of loss functions

A loss function (or objective function, or optimization score function) is one of the two parameters required to compile a model:

```
model.compile(loss='mean_squared_error', optimizer='sgd')

from keras import losses
model.compile(loss=losses.mean_squared_error, optimizer='sgd')
```

You can either pass the name of an existing loss function, or pass a TensorFlow/Theano symbolic function that returns a scalar for each data-point and takes the following two arguments:

1. y_true: True labels. TensorFlow/Theano tensor.
2. y_pred: Predictions. TensorFlow/Theano tensor of the same shape as y_true.

The actual optimized objective is the mean of the output array across all datapoints.

In [0]:
import keras
from keras import models
from keras.layers import Dense, Dropout
from keras.utils import to_categorical
from keras.datasets import mnist
from keras.utils.vis_utils import model_to_dot
from IPython.display import SVG

NUM_ROWS = 28
NUM_COLS = 28
NUM_CLASSES = 10
BATCH_SIZE = 128
EPOCHS = 10

# Load data
(X_train, y_train), (X_test, y_test) = mnist.load_data()

# Reshape data
X_train = X_train.reshape((X_train.shape[0], NUM_ROWS * NUM_COLS))
X_train = X_train.astype('float32') / 255
X_test = X_test.reshape((X_test.shape[0], NUM_ROWS * NUM_COLS))
X_test = X_test.astype('float32') / 255

# Categorically encode labels
y_train = to_categorical(y_train, NUM_CLASSES)
y_test = to_categorical(y_test, NUM_CLASSES)


# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='elu', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='elu'))
model.add(Dense(10, activation='softmax'))



Using TensorFlow backend.


11493376/11490434 [==============================] - 1s 0us/step





#mean_squared_error

keras.losses.mean_squared_error(y_true, y_pred)

In [0]:
# Compile model using accuracy
model.compile(optimizer='rmsprop',
              loss='mean_squared_error',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])





Train on 60000 samples, validate on 10000 samples
Epoch 1/10





60000/60000 [==============================] - 11s 178us/step - loss: 0.0158 - acc: 0.8941 - val_loss: 0.0081 - val_acc: 0.9454
Epoch 2/10
60000/60000 [==============================] - 1s 24us/step - loss: 0.0075 - acc: 0.9510 - val_loss: 0.0060 - val_acc: 0.9604
Epoch 3/10
60000/60000 [==============================] - 1s 24us/step - loss: 0.0053 - acc: 0.9657 - val_loss: 0.0050 - val_acc: 0.9666
Epoch 4/10
60000/60000 [==============================] - 1s 25us/step - loss: 0.0043 - acc: 0.9728 - val_loss: 0.0043 - val_acc: 0.9721
Epoch 5/10
60000/60000 [==============================] - 1s 24us/step - loss: 0.0036 - acc: 0.9768 - val_loss: 0.0050 - val_acc: 0.9700
Epoch 6/10
60000/60000 [==============================] - 2s 27us/step - loss: 0.0032 - acc: 0.9803 - val_loss: 0.0042 - val_acc: 0.9732
Epoch 7/10
60000/60000 [==============================] - 1s 24us/step - loss: 0.0027 - acc: 0.9830 - val_loss: 0.004

#mean_absolute_error

keras.losses.mean_absolute_error(y_true, y_pred)

In [0]:
# Compile model using accuracy
model.compile(optimizer='rmsprop',
              loss='mean_absolute_error',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 26us/step - loss: 0.0027 - acc: 0.9881 - val_loss: 0.0046 - val_acc: 0.9782
Epoch 2/10
60000/60000 [==============================] - 1s 24us/step - loss: 0.0022 - acc: 0.9901 - val_loss: 0.0044 - val_acc: 0.9790
Epoch 3/10
60000/60000 [==============================] - 1s 24us/step - loss: 0.0020 - acc: 0.9912 - val_loss: 0.0045 - val_acc: 0.9778
Epoch 4/10
60000/60000 [==============================] - 2s 26us/step - loss: 0.0018 - acc: 0.9918 - val_loss: 0.0044 - val_acc: 0.9789
Epoch 5/10
60000/60000 [==============================] - 1s 24us/step - loss: 0.0017 - acc: 0.9921 - val_loss: 0.0041 - val_acc: 0.9799
Epoch 6/10
60000/60000 [==============================] - 1s 24us/step - loss: 0.0015 - acc: 0.9927 - val_loss: 0.0047 - val_acc: 0.9770
Epoch 7/10
60000/60000 [==============================] - 2s 27us/step - loss: 0.0015 - acc: 0.9930 - val_loss: 0.0042 - val_acc

#mean_absolute_percentage_error

keras.losses.mean_absolute_percentage_error(y_true, y_pred)

In [0]:
# Compile model using accuracy
model.compile(optimizer='rmsprop',
              loss='mean_absolute_percentage_error',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 29us/step - loss: 622485.9289 - acc: 0.9940 - val_loss: 2138205.3277 - val_acc: 0.9789
Epoch 2/10
60000/60000 [==============================] - 1s 25us/step - loss: 585743.9731 - acc: 0.9944 - val_loss: 1925244.3143 - val_acc: 0.9810
Epoch 3/10
60000/60000 [==============================] - 2s 26us/step - loss: 589206.0974 - acc: 0.9944 - val_loss: 1922228.6152 - val_acc: 0.9811
Epoch 4/10
60000/60000 [==============================] - 2s 26us/step - loss: 552529.5558 - acc: 0.9948 - val_loss: 1925542.5419 - val_acc: 0.9811
Epoch 5/10
60000/60000 [==============================] - 2s 26us/step - loss: 560338.6004 - acc: 0.9946 - val_loss: 1821772.2996 - val_acc: 0.9817
Epoch 6/10
60000/60000 [==============================] - 2s 26us/step - loss: 521652.2927 - acc: 0.9950 - val_loss: 1999078.2744 - val_acc: 0.9801
Epoch 7/10
60000/60000 [==============================] - 1s 2

#mean_squared_logarithmic_error

keras.losses.mean_squared_logarithmic_error(y_true, y_pred)

In [0]:
# Compile model using accuracy
model.compile(optimizer='rmsprop',
              loss='mean_squared_logarithmic_error',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])


Instructions for updating:
Use tf.where in 2.0, which has the same broadcast rule as np.where
Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 30us/step - loss: 4.3306e-04 - acc: 0.9951 - val_loss: 0.0020 - val_acc: 0.9778
Epoch 2/10
60000/60000 [==============================] - 2s 27us/step - loss: 4.3759e-04 - acc: 0.9950 - val_loss: 0.0018 - val_acc: 0.9801
Epoch 3/10
60000/60000 [==============================] - 2s 27us/step - loss: 4.3795e-04 - acc: 0.9951 - val_loss: 0.0018 - val_acc: 0.9799
Epoch 4/10
60000/60000 [==============================] - 2s 26us/step - loss: 4.0980e-04 - acc: 0.9955 - val_loss: 0.0018 - val_acc: 0.9795
Epoch 5/10
60000/60000 [==============================] - 2s 27us/step - loss: 4.3924e-04 - acc: 0.9951 - val_loss: 0.0017 - val_acc: 0.9804
Epoch 6/10
60000/60000 [==============================] - 2s 27us/step - loss: 4.6374e-04 - acc: 0.9947 - val_loss: 0.0017 - val_acc: 0.9809
Epoch 7/1

#squared_hinge

keras.losses.squared_hinge(y_true, y_pred)

In [0]:
# Compile model using accuracy
model.compile(optimizer='rmsprop',
              loss='squared_hinge',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.9004 - acc: 0.9956 - val_loss: 0.9019 - val_acc: 0.9804
Epoch 2/10
60000/60000 [==============================] - 1s 25us/step - loss: 0.9004 - acc: 0.9960 - val_loss: 0.9019 - val_acc: 0.9791
Epoch 3/10
60000/60000 [==============================] - 2s 27us/step - loss: 0.9004 - acc: 0.9958 - val_loss: 0.9018 - val_acc: 0.9811
Epoch 4/10
60000/60000 [==============================] - 2s 25us/step - loss: 0.9004 - acc: 0.9960 - val_loss: 0.9018 - val_acc: 0.9816
Epoch 5/10
60000/60000 [==============================] - 2s 26us/step - loss: 0.9004 - acc: 0.9962 - val_loss: 0.9018 - val_acc: 0.9806
Epoch 6/10
60000/60000 [==============================] - 2s 26us/step - loss: 0.9003 - acc: 0.9964 - val_loss: 0.9018 - val_acc: 0.9805
Epoch 7/10
60000/60000 [==============================] - 1s 25us/step - loss: 0.9004 - acc: 0.9961 - val_loss: 0.9017 - val_acc

#hinge

keras.losses.hinge(y_true, y_pred)

In [0]:
# Compile model using accuracy
model.compile(optimizer='rmsprop',
              loss='hinge',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 31us/step - loss: 0.9003 - acc: 0.9967 - val_loss: 0.9019 - val_acc: 0.9811
Epoch 2/10
60000/60000 [==============================] - 1s 24us/step - loss: 0.9003 - acc: 0.9971 - val_loss: 0.9018 - val_acc: 0.9819
Epoch 3/10
60000/60000 [==============================] - 1s 25us/step - loss: 0.9003 - acc: 0.9968 - val_loss: 0.9018 - val_acc: 0.9818
Epoch 4/10
60000/60000 [==============================] - 2s 26us/step - loss: 0.9003 - acc: 0.9968 - val_loss: 0.9018 - val_acc: 0.9821
Epoch 5/10
60000/60000 [==============================] - 2s 25us/step - loss: 0.9003 - acc: 0.9970 - val_loss: 0.9018 - val_acc: 0.9817
Epoch 6/10
60000/60000 [==============================] - 2s 25us/step - loss: 0.9003 - acc: 0.9968 - val_loss: 0.9018 - val_acc: 0.9819
Epoch 7/10
60000/60000 [==============================] - 1s 25us/step - loss: 0.9003 - acc: 0.9970 - val_loss: 0.9020 - val_acc

#categorical_hinge

keras.losses.categorical_hinge(y_true, y_pred)

In [0]:
# Compile model using accuracy
model.compile(optimizer='rmsprop',
              loss='categorical_hinge',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 31us/step - loss: 0.0062 - acc: 0.9969 - val_loss: 0.0349 - val_acc: 0.9827
Epoch 2/10
60000/60000 [==============================] - 2s 27us/step - loss: 0.0061 - acc: 0.9970 - val_loss: 0.0372 - val_acc: 0.9814
Epoch 3/10
60000/60000 [==============================] - 2s 26us/step - loss: 0.0060 - acc: 0.9970 - val_loss: 0.0331 - val_acc: 0.9837
Epoch 4/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0059 - acc: 0.9971 - val_loss: 0.0369 - val_acc: 0.9818
Epoch 5/10
60000/60000 [==============================] - 2s 26us/step - loss: 0.0058 - acc: 0.9972 - val_loss: 0.0327 - val_acc: 0.9836
Epoch 6/10
60000/60000 [==============================] - 2s 26us/step - loss: 0.0061 - acc: 0.9970 - val_loss: 0.0364 - val_acc: 0.9819
Epoch 7/10
60000/60000 [==============================] - 2s 26us/step - loss: 0.0062 - acc: 0.9969 - val_loss: 0.0361 - val_acc

#logcosh

keras.losses.logcosh(y_true, y_pred)

Logarithm of the hyperbolic cosine of the prediction error.

log(cosh(x)) is approximately equal to (x ** 2) / 2 for small x and to abs(x) - log(2) for large x. This means that 'logcosh' works mostly like the mean squared error, but will not be so strongly affected by the occasional wildly incorrect prediction.

##Arguments

1. y_true: tensor of true targets.
2. y_pred: tensor of predicted targets.

##Returns

1. Tensor with one scalar loss entry per sample.

In [0]:
# Compile model using accuracy
model.compile(optimizer='rmsprop',
              loss='logcosh',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 31us/step - loss: 2.5007e-04 - acc: 0.9970 - val_loss: 0.0014 - val_acc: 0.9836
Epoch 2/10
60000/60000 [==============================] - 1s 25us/step - loss: 2.6314e-04 - acc: 0.9968 - val_loss: 0.0016 - val_acc: 0.9802
Epoch 3/10
60000/60000 [==============================] - 1s 25us/step - loss: 2.5415e-04 - acc: 0.9970 - val_loss: 0.0017 - val_acc: 0.9795
Epoch 4/10
60000/60000 [==============================] - 1s 25us/step - loss: 2.5471e-04 - acc: 0.9969 - val_loss: 0.0015 - val_acc: 0.9813
Epoch 5/10
60000/60000 [==============================] - 1s 25us/step - loss: 2.7108e-04 - acc: 0.9967 - val_loss: 0.0015 - val_acc: 0.9821
Epoch 6/10
60000/60000 [==============================] - 1s 24us/step - loss: 2.2052e-04 - acc: 0.9973 - val_loss: 0.0014 - val_acc: 0.9824
Epoch 7/10
60000/60000 [==============================] - 2s 26us/step - loss: 2.5697e-04 - acc: 0.9969 

#categorical_crossentropy

keras.losses.categorical_crossentropy(y_true, y_pred, from_logits=False, label_smoothing=0)

In [0]:
# Compile model using accuracy
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 34us/step - loss: 0.0369 - acc: 0.9972 - val_loss: 0.1890 - val_acc: 0.9833
Epoch 2/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.0369 - acc: 0.9972 - val_loss: 0.2055 - val_acc: 0.9828
Epoch 3/10
60000/60000 [==============================] - 2s 26us/step - loss: 0.0365 - acc: 0.9974 - val_loss: 0.1951 - val_acc: 0.9827
Epoch 4/10
60000/60000 [==============================] - 2s 27us/step - loss: 0.0364 - acc: 0.9974 - val_loss: 0.1995 - val_acc: 0.9822
Epoch 5/10
60000/60000 [==============================] - 2s 27us/step - loss: 0.0361 - acc: 0.9973 - val_loss: 0.1841 - val_acc: 0.9838
Epoch 6/10
60000/60000 [==============================] - 2s 26us/step - loss: 0.0348 - acc: 0.9973 - val_loss: 0.1860 - val_acc: 0.9831
Epoch 7/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0334 - acc: 0.9977 - val_loss: 0.2097 - val_acc

#binary_crossentropy

keras.losses.binary_crossentropy(y_true, y_pred, from_logits=False, label_smoothing=0)

In [0]:
# Compile model using accuracy
model.compile(optimizer='rmsprop',
              loss='binary_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 35us/step - loss: 0.0069 - acc: 0.9994 - val_loss: 0.0403 - val_acc: 0.9964
Epoch 2/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.0067 - acc: 0.9995 - val_loss: 0.0435 - val_acc: 0.9962
Epoch 3/10
60000/60000 [==============================] - 2s 27us/step - loss: 0.0069 - acc: 0.9994 - val_loss: 0.0411 - val_acc: 0.9963
Epoch 4/10
60000/60000 [==============================] - 2s 27us/step - loss: 0.0065 - acc: 0.9995 - val_loss: 0.0393 - val_acc: 0.9963
Epoch 5/10
60000/60000 [==============================] - 2s 26us/step - loss: 0.0066 - acc: 0.9994 - val_loss: 0.0400 - val_acc: 0.9964
Epoch 6/10
60000/60000 [==============================] - 2s 27us/step - loss: 0.0065 - acc: 0.9995 - val_loss: 0.0365 - val_acc: 0.9965
Epoch 7/10
60000/60000 [==============================] - 2s 27us/step - loss: 0.0066 - acc: 0.9994 - val_loss: 0.0381 - val_acc

#kullback_leibler_divergence

keras.losses.kullback_leibler_divergence(y_true, y_pred)

In [0]:
# Compile model using accuracy
model.compile(optimizer='rmsprop',
              loss='kullback_leibler_divergence',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 33us/step - loss: 0.0315 - acc: 0.9976 - val_loss: 0.2007 - val_acc: 0.9824
Epoch 2/10
60000/60000 [==============================] - 2s 25us/step - loss: 0.0304 - acc: 0.9976 - val_loss: 0.1945 - val_acc: 0.9833
Epoch 3/10
60000/60000 [==============================] - 2s 25us/step - loss: 0.0296 - acc: 0.9979 - val_loss: 0.1889 - val_acc: 0.9832
Epoch 4/10
60000/60000 [==============================] - 2s 25us/step - loss: 0.0287 - acc: 0.9979 - val_loss: 0.1979 - val_acc: 0.9827
Epoch 5/10
60000/60000 [==============================] - 1s 25us/step - loss: 0.0277 - acc: 0.9980 - val_loss: 0.2069 - val_acc: 0.9820
Epoch 6/10
60000/60000 [==============================] - 2s 25us/step - loss: 0.0271 - acc: 0.9981 - val_loss: 0.2080 - val_acc: 0.9817
Epoch 7/10
60000/60000 [==============================] - 2s 25us/step - loss: 0.0262 - acc: 0.9982 - val_loss: 0.1966 - val_acc

#poisson

keras.losses.poisson(y_true, y_pred)

In [0]:
# Compile model using accuracy
model.compile(optimizer='rmsprop',
              loss='poisson',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 32us/step - loss: 0.1028 - acc: 0.9980 - val_loss: 0.1211 - val_acc: 0.9832
Epoch 2/10
60000/60000 [==============================] - 1s 25us/step - loss: 0.1026 - acc: 0.9982 - val_loss: 0.1228 - val_acc: 0.9812
Epoch 3/10
60000/60000 [==============================] - 1s 25us/step - loss: 0.1027 - acc: 0.9981 - val_loss: 0.1222 - val_acc: 0.9811
Epoch 4/10
60000/60000 [==============================] - 1s 24us/step - loss: 0.1027 - acc: 0.9982 - val_loss: 0.1233 - val_acc: 0.9800
Epoch 5/10
60000/60000 [==============================] - 1s 24us/step - loss: 0.1026 - acc: 0.9983 - val_loss: 0.1223 - val_acc: 0.9811
Epoch 6/10
60000/60000 [==============================] - 2s 27us/step - loss: 0.1026 - acc: 0.9982 - val_loss: 0.1227 - val_acc: 0.9811
Epoch 7/10
60000/60000 [==============================] - 1s 25us/step - loss: 0.1026 - acc: 0.9983 - val_loss: 0.1227 - val_acc

#cosine_proximity

keras.losses.cosine_proximity(y_true, y_pred, axis=-1)

In [0]:
# Compile model using accuracy
model.compile(optimizer='rmsprop',
              loss='cosine_proximity',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 36us/step - loss: -0.9979 - acc: 0.9979 - val_loss: -0.9815 - val_acc: 0.9813
Epoch 2/10
60000/60000 [==============================] - 2s 27us/step - loss: -0.9981 - acc: 0.9980 - val_loss: -0.9819 - val_acc: 0.9815
Epoch 3/10
60000/60000 [==============================] - 2s 30us/step - loss: -0.9982 - acc: 0.9982 - val_loss: -0.9807 - val_acc: 0.9804
Epoch 4/10
60000/60000 [==============================] - 2s 26us/step - loss: -0.9981 - acc: 0.9980 - val_loss: -0.9816 - val_acc: 0.9813
Epoch 5/10
60000/60000 [==============================] - 2s 27us/step - loss: -0.9981 - acc: 0.9981 - val_loss: -0.9805 - val_acc: 0.9798
Epoch 6/10
60000/60000 [==============================] - 2s 29us/step - loss: -0.9981 - acc: 0.9980 - val_loss: -0.9835 - val_acc: 0.9832
Epoch 7/10
60000/60000 [==============================] - 2s 27us/step - loss: -0.9982 - acc: 0.9982 - val_loss: -0.

#Thank you for completing this notebook

1)**mean_squared_error**

Test loss: 0.003352416222123199


Test accuracy: 0.9799

2)**mean_absolute_error**

Test loss: 0.0038889680898445095


Test accuracy: 0.981

3)**mean_absolute_percentage_error**

Test loss: 1803200.477230715


Test accuracy: 0.9819

4)**mean_squared_logarithmic_error**

Test loss: 0.0017203592830507266


Test accuracy: 0.9807

5)**squared_hinge**

Test loss: 0.9018609205245972


Test accuracy: 0.9803

6)**hinge**

Test loss: 0.901724101448059


Test accuracy: 0.9833

7)**categorical_hinge**

Test loss: 0.03607132461245253


Test accuracy: 0.982

8)**logcosh**

Test loss: 0.0015778261381334592


Test accuracy: 0.981

9)**categorical_crossentropy**

Test loss: 0.23053656777290435


Test accuracy: 0.9802

10)**binary_crossentropy**

Test loss: 0.04225646639997729


Test accuracy: 0.996229995727539

11)**kullback_leibler_divergence**

Test loss: 0.21674129924554209


Test accuracy: 0.9822

12)**poisson**

Test loss: 0.1220341134428978


Test accuracy: 0.9826

13)**cosine_proximity**

Test loss: -0.9826041687011718


Test accuracy: 0.9819

The best among all is **binary_crossentropy** with an accuracy of almost 1.

In [0]:
from keras.datasets import cifar10
from keras.models import Sequential
from keras.layers import Dense, Dropout, Activation, Flatten
from keras.utils import to_categorical

#import dataset
(X_train, Y_train), (X_test, Y_test) = cifar10.load_data()

#change shape from image to vector
X_train = X_train.reshape(50000, 32 * 32 * 3)
X_test = X_test.reshape(10000, 32 * 32 * 3)

#preprocess
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
X_train /= 255.0
X_test /= 255.0

#change labels from numeric to one hot encoded
Y_train = to_categorical(Y_train, 10)
Y_test =  to_categorical(Y_test, 10)

#Model building
model = Sequential()
model.add(Dense(1024, input_shape=(3072, )))
model.add(Activation('relu'))
model.add(Dense(512))
model.add(Activation('relu'))
model.add(Dense(512))
model.add(Activation('relu'))
model.add(Dense(10))
model.add(Activation('softmax'))

# Compile model using above optimizer
model.compile(optimizer='rmsprop',
              loss='binary_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, Y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, Y_test))

score = model.evaluate(X_test, Y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

170500096/170498071 [==============================] - 4s 0us/step
Train on 50000 samples, validate on 10000 samples
Epoch 1/10
50000/50000 [==============================] - 3s 54us/step - loss: 0.4940 - acc: 0.8926 - val_loss: 0.2760 - val_acc: 0.9026
Epoch 2/10
50000/50000 [==============================] - 2s 42us/step - loss: 0.2710 - acc: 0.9021 - val_loss: 0.2647 - val_acc: 0.9027
Epoch 3/10
50000/50000 [==============================] - 2s 41us/step - loss: 0.2571 - acc: 0.9048 - val_loss: 0.2505 - val_acc: 0.9069
Epoch 4/10
50000/50000 [==============================] - 2s 41us/step - loss: 0.2481 - acc: 0.9072 - val_loss: 0.2511 - val_acc: 0.9063
Epoch 5/10
50000/50000 [==============================] - 2s 42us/step - loss: 0.2410 - acc: 0.9088 - val_loss: 0.2491 - val_acc: 0.9070
Epoch 6/10
50000/50000 [==============================] - 2s 41us/step - loss: 0.2353 - acc: 0.9107 - val_loss: 0.2444 - val_acc: 0.9084
Epoch 7/10
50000/50000 [==============================] - 2s 

**CIFAR 10**

Test loss: 0.23143027470111846


Test accuracy: 0.9121199871063232

In [0]:
from keras.datasets import cifar100
from keras.models import Sequential
from keras.layers import Dense, Dropout, Activation, Flatten
from keras.utils import to_categorical

#import dataset
(X_train, Y_train), (X_test, Y_test) = cifar100.load_data()

#change shape from image to vector
X_train = X_train.reshape(50000, 32 * 32 * 3)
X_test = X_test.reshape(10000, 32 * 32 * 3)

#preprocess
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
X_train /= 255.0
X_test /= 255.0

#change labels from numeric to one hot encoded
Y_train = to_categorical(Y_train, 100)
Y_test =  to_categorical(Y_test, 100)

#Model building
model = Sequential()
model.add(Dense(1024, input_shape=(3072, )))
model.add(Activation('relu'))
model.add(Dense(512))
model.add(Activation('relu'))
model.add(Dense(512))
model.add(Activation('relu'))
model.add(Dense(100))
model.add(Activation('softmax'))


from keras import optimizers
adamax=keras.optimizers.Adamax(lr=0.002, beta_1=0.9, beta_2=0.999)

# Compile model using above optimizer
model.compile(optimizer='rmsprop',
              loss='binary_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, Y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, Y_test))

score = model.evaluate(X_test, Y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

169009152/169001437 [==============================] - 4s 0us/step
Train on 50000 samples, validate on 10000 samples
Epoch 1/10
50000/50000 [==============================] - 3s 57us/step - loss: 0.0562 - acc: 0.9897 - val_loss: 0.0495 - val_acc: 0.9900
Epoch 2/10
50000/50000 [==============================] - 2s 44us/step - loss: 0.0477 - acc: 0.9900 - val_loss: 0.0474 - val_acc: 0.9900
Epoch 3/10
50000/50000 [==============================] - 2s 42us/step - loss: 0.0457 - acc: 0.9901 - val_loss: 0.0454 - val_acc: 0.9901
Epoch 4/10
50000/50000 [==============================] - 2s 43us/step - loss: 0.0443 - acc: 0.9901 - val_loss: 0.0440 - val_acc: 0.9902
Epoch 5/10
50000/50000 [==============================] - 2s 42us/step - loss: 0.0431 - acc: 0.9902 - val_loss: 0.0437 - val_acc: 0.9902
Epoch 6/10
50000/50000 [==============================] - 2s 44us/step - loss: 0.0421 - acc: 0.9902 - val_loss: 0.0449 - val_acc: 0.9902
Epoch 7/10
50000/50000 [==============================] - 2s 

**CIFAR 100**

Test loss: 0.04387838685512543


Test accuracy: 0.9902030138015747

In [0]:
import keras
from keras import models
from keras.layers import Dense, Dropout
from keras.utils import to_categorical
from keras.datasets import fashion_mnist
from keras.utils.vis_utils import model_to_dot
from IPython.display import SVG

NUM_ROWS = 28
NUM_COLS = 28
NUM_CLASSES = 10
BATCH_SIZE = 128
EPOCHS = 10

# Load data
(X_train, Y_train), (X_test, Y_test) = fashion_mnist.load_data()



# Reshape data
X_train = X_train.reshape((X_train.shape[0], NUM_ROWS * NUM_COLS))
X_train = X_train.astype('float32') / 255
X_test = X_test.reshape((X_test.shape[0], NUM_ROWS * NUM_COLS))
X_test = X_test.astype('float32') / 255

# Categorically encode labels
Y_train = to_categorical(Y_train, NUM_CLASSES)
Y_test = to_categorical(Y_test, NUM_CLASSES)


# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='relu', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='relu'))
model.add(Dense(10, activation='softmax'))

# Compile model using above optimizer
model.compile(optimizer='rmsprop',
              loss='binary_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, Y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, Y_test))

score = model.evaluate(X_test, Y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

4423680/4422102 [==============================] - 1s 0us/step
Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 39us/step - loss: 0.0937 - acc: 0.9621 - val_loss: 0.0787 - val_acc: 0.9678
Epoch 2/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0668 - acc: 0.9731 - val_loss: 0.0788 - val_acc: 0.9686
Epoch 3/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.0599 - acc: 0.9759 - val_loss: 0.0619 - val_acc: 0.9756
Epoch 4/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0557 - acc: 0.9778 - val_loss: 0.0646 - val_acc: 0.9754
Epoch 5/10
60000/60000 [==============================] - 2s 27us/step - loss: 0.0533 - acc: 0.9787 - val_loss: 0.0648 - val_acc: 0.9743
Epoch 6/10
60000/60000 [==============================] - 2s 27us/step - loss: 0.0505 - acc: 0.9798 - val_loss: 0.0669 - val_acc: 0.9747
Epoch 7/10
60000/60000 [==============================] - 2s 28us

**FASHION_MNIST**

Test loss: 0.06269588343054056


Test accuracy: 0.9776000038146972